parser/parser.go

```diff
@@ -149,17 +149,27 @@ func Pin(ctx context.Context, res resolver.Resolver, parser Parser, nodes []*yam
 				return
 			}
 
-			resolved, err := res.Resolve(ctx, ref)
+			newRef := ref
+			if upgrade {
```

Comment:
I don't love this as an API, but I realize that it'd be annoying to duplicate 90% of this just for the upgrade path.